tests/test_utils.py

```diff
@@ -1510,3 +1510,105 @@ def test_use_sto4_return_true(client: flask.testing.FlaskClient):
     # Run function
     result: bool = use_sto4(unit_object=unit, project_object=project)
     assert result is True
+
+
+def test_add_uploaded_files_to_db_correct_failed_op(client: flask.testing.FlaskClient):
+    # Mock input data
+    proj_in_db = models.Project.query.first()
+    log = {
+        "file1.txt": {
+            "status": {"failed_op": "add_file_db"},
+            "path_remote": "path/to/file1.txt",
+            "subpath": "subpath",
+            "size_raw": 100,
+            "size_processed": 200,
+            "compressed": False,
+            "public_key": "public_key",
+            "salt": "salt",
+            "checksum": "checksum",
+        }
+    }
+
+    # Mock the S3 connector and head_object method
+    mock_s3conn = MagicMock()
+    mock_s3conn.resource.meta.client.head_object.return_value = None
+
+    # Call the function
+    with patch("dds_web.api.api_s3_connector.ApiS3Connector", return_value=mock_s3conn):
+        utils.add_uploaded_files_to_db(proj_in_db, log)
+
+    # check that the file is added to the database
+    file = models.File.query.filter_by(name="file1.txt").first()
+    assert file
+    assert file.name == "file1.txt"
+    assert file.name_in_bucket == "path/to/file1.txt"
+
+    # check that the file is added to the project
+    assert file in proj_in_db.files
+
+    # check that the version is added to the database
+    version = models.Version.query.filter_by(active_file=file.id).first()
+    assert version
+
+
+def test_add_uploaded_files_to_db_other_failed_op(client: flask.testing.FlaskClient):
+    # Mock input data
+    proj_in_db = models.Project.query.first()
+    log = {
+        "file1.txt": {
+            "status": {"failed_op": "some_other_failed_op"},
+            "path_remote": "path/to/file1.txt",
+            "subpath": "subpath",
+            "size_raw": 100,
+            "size_processed": 200,
+            "compressed": False,
+            "public_key": "public_key",
+            "salt": "salt",
+            "checksum": "checksum",
+        }
+    }
+
+    # Mock the S3 connector and head_object method
+    mock_s3conn = MagicMock()
+    mock_s3conn.resource.meta.client.head_object.return_value = None
+
+    # Call the function
+    with patch("dds_web.api.api_s3_connector.ApiS3Connector", return_value=mock_s3conn):
+        utils.add_uploaded_files_to_db(proj_in_db, log)
+
+    # check that the file is added to the database
+    file = models.File.query.filter_by(name="file1.txt").first()
+    assert not file
+
+
+# def test_add_uploaded_files_to_db_file_not_found(client: flask.testing.FlaskClient):
```

Comment:
I have been playing around, try this. it works on my enviroment

```
def test_add_uploaded_files_to_db_file_not_found(client: flask.testing.FlaskClient, capfd):

    from botocore.exceptions import ClientError
    # create mock project and log
    proj_in_db = models.Project.query.first()
    log = {
         "file1.txt": {
             "status": {"failed_op": "add_file_db"},
             "path_remote": "path/to/file1.txt",
             "subpath": "subpath",
             "size_raw": 100,
             "size_processed": 200,
             "compressed": False,
             "public_key": "public_key",
             "salt": "salt",
             "checksum": "checksum",
         }
    }

    mock_api_s3_conn = MagicMock()
    mock_s3conn = mock_api_s3_conn.return_value.__enter__.return_value
    with patch("dds_web.api.api_s3_connector.ApiS3Connector", mock_api_s3_conn):
        mock_s3conn.resource.meta.client.head_object.side_effect = ClientError({"Error": {"Code": "404"}}, "operation_name")
        utils.add_uploaded_files_to_db(proj_in_db, log)

    file = models.File.query.filter_by(name="file1.txt").first()
    assert not file
    _,err = capfd.readouterr()
    assert "Errors while adding files: {'file1.txt': {'error': 'File not found in S3" in err
```

Reply:
Thank you! 
It works in general, but I don't get anything returned by 'capfd.readouterr()' and the last assert fails with 
'assert "Errors while adding files: {'file1.txt': {'error': 'File not found in S3" in '''

Reply:
'capfd' didn't work because the errors are not actually printed out, they are returned by the function, instead. I have added checks for the return values now.